// @webpack-cli/generators/lib/addon-generator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const yeoman_generator_1 = __importDefault(require("yeoman-generator"));
const helpers_1 = require("./utils/helpers");
// Helper to get the template-directory content
const getFiles = (dir) => {
    return fs_1.default.readdirSync(dir).reduce((list, file) => {
        const filePath = path_1.default.join(dir, file);
        const isDir = fs_1.default.statSync(filePath).isDirectory();
        return list.concat(isDir ? getFiles(filePath) : filePath);
    }, []);
};
/**
 * Creates a Yeoman Generator that generates a project conforming
 * to webpack-defaults.
 *
 * @param {Generator.Questions} prompts An array of Yeoman prompt objects
 *
 * @param {string} templateDir Absolute path to template directory
 *
 * @param {Function} templateFn A function that is passed a generator instance and
 * returns an object containing data to be supplied to the template files.
 *
 * @returns {Generator} A class extending Generator
 */
const addonGenerator = (prompts, templateDir, 
// eslint-disable-next-line @typescript-eslint/no-explicit-any
templateFn) => {
    return class extends yeoman_generator_1.default {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        constructor(args, opts) {
            super(args, opts);
            const { cli = {}, options } = opts || {};
            this.cli = cli;
            this.template = options.template;
            this.supportedTemplates = fs_1.default.readdirSync(templateDir);
        }
        async prompting() {
            this.template = await helpers_1.getTemplate.call(this);
            this.resolvedTemplatePath = path_1.default.join(templateDir, this.template);
            this.props = await this.prompt(prompts);
            this.packageManager = await helpers_1.getInstaller.call(this);
        }
        default() {
            const currentDirName = path_1.default.basename(this.destinationPath());
            if (currentDirName !== this.props.name) {
                this.log(`
				Your project must be inside a folder named ${this.props.name}
				I will create this folder for you.
                `);
                const pathToProjectDir = this.destinationPath(this.props.name);
                try {
                    fs_1.default.mkdirSync(pathToProjectDir, { recursive: true });
                }
                catch (error) {
                    this.cli.logger.error("Failed to create directory");
                    this.cli.logger.error(error);
                }
                this.destinationRoot(pathToProjectDir);
            }
        }
        writing() {
            const packageJsonTemplatePath = "../addon-template/package.json.js";
            this.fs.extendJSON(this.destinationPath("package.json"), 
            // eslint-disable-next-line @typescript-eslint/no-var-requires
            require(packageJsonTemplatePath)(this.props.name));
            let files = [];
            try {
                // An array of file paths (relative to `./templates`) of files to be copied to the generated project
                files = getFiles(this.resolvedTemplatePath);
            }
            catch (error) {
                this.cli.logger.error(`Failed to generate starter template.\n ${error}`);
                process.exit(2);
            }
            // Template file paths should be of the form `path/to/_file.js.tpl`
            const copyTemplateFiles = files.filter((filePath) => path_1.default.basename(filePath).startsWith("_"));
            // File paths should be of the form `path/to/file.js.tpl`
            const copyFiles = files.filter((filePath) => !copyTemplateFiles.includes(filePath));
            copyFiles.forEach((filePath) => {
                // `absolute-path/to/file.js.tpl` -> `destination-path/file.js`
                const destFilePath = path_1.default.relative(this.resolvedTemplatePath, filePath).replace(".tpl", "");
                this.fs.copyTpl(filePath, this.destinationPath(destFilePath));
            });
            copyTemplateFiles.forEach((filePath) => {
                // `absolute-path/to/_file.js.tpl` -> `destination-path/file.js`
                const destFilePath = path_1.default
                    .relative(this.resolvedTemplatePath, filePath)
                    .replace("_", "")
                    .replace(".tpl", "");
                this.fs.copyTpl(filePath, this.destinationPath(destFilePath), templateFn(this));
            });
        }
        install() {
            const opts = this.packageManager === "yarn" ? { dev: true } : { "save-dev": true };
            this.scheduleInstallTask(this.packageManager, ["webpack-defaults", "bluebird"], opts);
        }
    };
};
exports.default = addonGenerator;
